csharp/src/Ice/OutgoingRequestFrame.cs

```diff
@@ -195,14 +224,19 @@ public sealed class OutgoingRequestFrame : OutgoingFrame
             IsSealed = Protocol == Protocol.Ice1;
         }
 
-        // Finish prepare the frame for sending, write the frame Context into the slot 0 of the binary context.
+        // Finish prepares the frame for sending ad writes the frame's context into slot 0 of the binary context.
         internal override void Finish()
         {
             if (!IsSealed)
             {
-                if (Protocol == Protocol.Ice2 && Context.Count > 0 && !_binaryContextKeys.Contains(0))
+                if (Protocol == Protocol.Ice2 && !ContainsKey(0))
                 {
-                    AddBinaryContextEntry(0, Context, ContextHelper.IceWriter);
+                    if (Context.Count > 0 || _writeSlot0)
+                    {
+                        // When _writeSlot0 is true, we may write an empty string-string context, thus preventing base
+                        // from writing a non-empty Context.
+                        AddBinaryContextEntry(0, Context, ContextHelper.IceWriter);
```

Comment:
I don't understand this, is it to forward the incoming empty context and ignore overrides?

Reply:
The situation we avoid here is when _defaultBinaryContext has a non-empty slot 0 - meaning a non-empty string-string Context - while the application set Context to be empty. In this case, if we skip slot 0 to mean "empty context", OutgoingFrame.Finish would write the non-empty slot 0, which is not what we want.

_writeSlot0 == true avoids this problem by forcing the writing of slot 0, even when Context is empty.

Reply:
I see it makes sense now